README.md
# OSBuild - FOR ANDROID DEVELOPERS

![icon.png](/Other/icon.png)

Ever wanted to peek inside and know more about your device? Well, **OSBuild**, a simple, lightweight, clean and beautiful app, meets your needs. And it works on all phones running Android 2.2 and above!

A simple UI and some nice features make it great tool especially for Android developers and learners.

### Features
* List **RICH RAW** info about the device with keeping most programming words.
* Provide **RICH** related data, like `19  KITKAT  4.4  2013-10  39.3%  19`.
* List all features the system supported, such as `FEATURE_USB_HOST`.
* Expose details of installed apps, including **SIGNATURE**.
* **OPEN SOURCE**. You can see its source codes on [GitHub](https://github.com/by-syk/OSBuild "GitHub") to get useful code sections.

### Modules
* Primer [the outline for average uses]
* Build [`MODEL`, etc.]
* Display [`densityDpi`, etc.]
* Telephony [`getDeviceId()`, etc.]
* CPU [`cpuinfo_max_freq`, etc.]
* Memory [`getVolumePaths()`, etc.]
* Package [`getSystemAvailableFeatures()`, etc.]
* Sensor [`getSensorList(TYPE_ALL)`, etc.]
* Superuser [`BusyBox`, etc.]
* App [`requestedPermissions`, etc.]
* etc.

### Example
There is a slice of data which **OSBuild** got from my Android phone:
```
/proc/cpuinfo
   Processor: ARMv7 Processor rev 1 (v7l)
   Hardware: Qualcomm MSM8974PRO-AB
/sys/devices/system/cpu/
   online: 0,2-3  3
   cpu[0-9]/  4
   cpu0/cpufreq/
      cpuinfo_min_freq: 300000  300.00MHz
      cpuinfo_max_freq: 2265600  2.27GHz
      scaling_governor: interactive
javax.microedition.khronos.opengles.GL10.
   glGetString(GL_RENDERER): Adreno (TM) 330
   glGetString(GL_VENDOR): Qualcomm
/sys/class/kgsl/kgsl-3d0/
   max_gpuclk: 578000000  578.00MHz
```

### Raise Your Voice
Any BUGs, mistakes, suggestion or confusion, feedback by [E-mail](mailto:By_syk@163.com "Developer's E-mail address") or revise on a market. I would take to heart.

Your voice can make a difference!

### Donate
Not financially.

Just **Share the Page** to me by E-mail or any other available way. As you can see, only some hardware info about the device (**NO PRIVACY**).

Each received file helps me expand the **OSBuild** - **DAD (Database of Android Devices)**.

I've spent plenty of time programming. If you find **OSBuild** useful, please help me a lot. Thank you.

### Besides...
Maybe **OSBuild** could help you with smartphone reviews.

Salute to [Sony Xperia AppXplore](http://developer.sonymobile.com/knowledge-base/tools/download-appxplore-and-learn-more-about-your-apps/ "AppXplore").

### History Versions
[OSBuild_v1.1.0(150606).apk](/Other/OSBuild_v1.1.0(150606).apk)

[OSBuild_v1.1.4(150610).apk](/Other/OSBuild_v1.1.4(150610).apk)

[OSBuild_v1.2.0(150628).apk](/Other/OSBuild_v1.2.0(150628).apk)

[OSBuild_v1.2.7(150708).apk](/Other/OSBuild_v1.2.7(150708).apk)

[OSBuild_v1.3.0(150723).apk](/Other/OSBuild_v1.3.0(150723).apk)

[OSBuild_v1.4.0(150830).apk](/Other/OSBuild_v1.4.0(150830).apk)

[OSBuild_v1.4.3(150911).apk](/Other/OSBuild_v1.4.3(150911).apk)

[OSBuild_v1.4.7(150928).apk](/Other/OSBuild_v1.4.7(150928).apk)

[OSBuild_v1.4.9(151115).apk](/Other/OSBuild_v1.4.9(151115).apk)

### Screenshots
![screenshot.png](/Other/screenshot.png)

CHANGELOG.txt
v1.4.9.nightly 2015-11-15
Remove the node "devices/fdb00000.qcom,kgsl-3d0/kgsl/kgsl-3d0/devfreq/" for "MSM8974PRO" CPU in CPU Module.
Check box of "Don't show again." is always visible in dialogs.
Add extra info for more devices (690+).

v1.4.9.nightly 2015-11-06
Update Google Install Stats in "Constant Reference".
Add extra info for more devices (670+).

v1.4.8.nightly 2015-10-24
Add some data of release date after the line "SDK_INT" in Build Module.

v1.4.8.nightly 2015-10-12
Add the node "devices/fdb00000.qcom,kgsl-3d0/kgsl/kgsl-3d0/devfreq/" for "MSM8974PRO" CPU in CPU Module.

v1.4.8.nightly 2015-10-07
Update Google Install Stats in "Constant Reference".

v1.4.8.nightly 2015-10-03
Fix the BUG: it crashes when launching OSBuuld Reader with NULL URI.
Fix the BUG: it crashes if the content of text file is empty in OSBuild Reader.

v1.4.7.nightly 2015-09-26
Fix the BUG: it crashes if the signatures is null in App Module.

v1.4.7.nightly 2015-09-17
Add the line "minSdkVersion" in App Module.
OSBuild Reader could read AndroidManifest.xml of .apk file.

v1.4.6.nightly 2015-09-15
Remember what was selected last time in "Installed" of "Check App".
Add the node "PackageManager" and its line "getComponent()" in Primer Module.

v1.4.5.nightly 2015-09-14
Fix the issue: wrong method to analyse signatures info in the line "getIssuerDN()" (like "Google, Inc" to "Google" and "Inc").
Add extra info for more devices (500+).
Add unit "mm" after the line "Dimensions" in Primer Module.
Show lightness value in White Paper Activity.

v1.4.4.nightly 2015-09-13
Hide the line "flags & FLAG_SYSTEM" when the app is not installed in App Module.

v1.4.4.nightly 2015-09-12
Add the node "/sys/class/kgsl/kgsl-3d0/" in CPU Module.
Show file name as a subtitle on the ActionBar in OSBuild Reader.

v1.4.3.nightly 2015-09-11
Tap ENTER to perform related operation instead of just closing the soft keyboard in dialogs.
Add the lines "CPU_ABI" and "CPU_ABI2" in Build Module (API 21-).
Fix the issue: wrong method to convert unit of frequency. Fortunately, the result is correct still.
Optimize the launcher icon for mdpi devices.

v1.4.3.nightly 2015-09-10
Click hard search button (if there is one) to open the dialog "Check App" quickly.
Add the VM type ("Dalvik" or "ART") after the line "getProperty("java.vm.version")" in Build Module.

v1.4.2.nightly 2015-09-08
Fix the issue: it can't load extra info, execute "Share the Page" and "Share OSBuild" due to full external storage.
Enlarge the database of extra info (up to 41 brands, 276 models).

v1.4.2.nightly 2015-09-07
Fix the BUG: it crashes because database which saves extra info is not loaded.

v1.4.2.nightly 2015-09-06
Add the line "Pixel & Screen Area" in Primer Module.
Fix the BUG: wrong value in the line "cpuinfo_min_freq" in CPU Module.

v1.4.2.nightly 2015-09-05
Add a real value (not calculated) after the line "Diagonal Size" in Display Module.
Add the line "Dimensions" and "Weight" after the node "Extra" in Display Module.
Add a value (diagonal size) after the line "Screen Resolution" in Primer Module.
Add the line "Dimensions" in Primer Module.

v1.4.1.nightly 2015-09-03
Show modules one by one more smoothly when launching.

v1.4.1.nightly 2015-09-01
Play animation when showing the FAB.
The steps to get GPU info become more simple and controllable.

v1.4.1.nightly 2015-08-31
Modify the size of FAB icon from 36dp to 24dp.

v1.4.0.nightly 2015-08-30
Elimilate the risk: it crashes when clicking the FAB occasionally.
Fix the BUG: It crashed when get telephony info due to "permission denied" in API 23+.
Request for permission in API 23+.
Fix the issue: the style of dialog "What's OSBuild?" is different.

v1.4.0.nightly 2015-08-29
Remember text size in OSBuild Reader.
Fix the BUG: It crashes when open App Module by apk path on API 13- (because there is a BUG of Android 2.2 to get signature info).

v1.4.0.nightly 2015-08-28
Fix the BUG: It crashes when open App Module by apk path on Android 2.2 (because there is a BUG of Android 2.2 to get signature info).

v1.4.0.nightly 2015-08-27
Add lines "getNotBefore()", "getNotAfter()" and "getSigAlgName()" in App Module.
Modify the line "getSubjectDN()" to "getIssuerDN()" in App Module.

v1.4.0.nightly 2015-08-26
Fix lost "GL_ES_VERSION_UNDEFINED" issue after the line "reqGlEsVersion" in Memory Module.
Add the line "isLowRamDevice()" in Memory Module.
Now OSBuild can check app by its apk path.

v1.4.0.nightly 2015-08-25
Add the line "targetSdkVersion" in App Module.
Fix the BUG: It crashed when get telephony info due to "permission denied" in API 23+.
Add the line "World Phone" in Primer Module (API 23+).
Add lines "getPhoneCount()", "getDeviceId(0)", "getDeviceId(1)" and "isWorldPhone()" in Display Module (API 23+).
Add the line "getTargetApi()" in App Module.

v1.4.0.nightly 2015-08-24
Adjust the length of dot line to the width of the page.

v1.4.0.nightly 2015-08-23
Fix the issue: the dialog "What's OSBuild?" do not show when launching firstly.
Add lines "Hi-Fi", "Fingerprint ID" in Primer Module.

v1.4.0.nightly 2015-08-22
Fix the issue: the soft keyboard is still showing when open list of charsets dialog in OSBuild Reader.
Remember what was searched last time in "Check Installed App".

v1.4.0.nightly 2015-08-21
Fix the BUG: OSBuild crashes when reading CPU info in some devices.

v1.4.0.nightly 2015-08-20
Remove menu options: "LogCat", "Rotation Vector"， "Compass" and "PPI Calculator" in menu "Experimentally".
Add color name in the cornor of "White Paper" layout.
Fix the issue: indeterminate progress doesn't show on "OSBuild Reader".
Set self-defined charset in "OSBuild Reader".
Listening scale gesture to scale the text size (only 3: Small, Normal, Large) in "OSBuild Reader".
Show experimental features as default (show the memu option "Experimentally"). So we need not long press button "OK" in dialog "Donate" to unlock it.
Fix the issue: two same dialogs "What's OSBuild?" may show at the same time when launching firstly.

app/src/main/java/com/by_syk/osbuild/kube/Kube.java
/*
 * Copyright (C) 2008 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.by_syk.osbuild.kube;

import android.app.Activity;
import android.os.Bundle;
import android.view.Window;
import android.opengl.GLSurfaceView;
import java.util.Random;
import android.view.WindowManager;

public class Kube extends Activity implements KubeRenderer.AnimationCallback
{
    private GLWorld makeGLWorld()
    {
        GLWorld world = new GLWorld();

        int one = 0x10000;
        int half = 0x08000;
        GLColor red = new GLColor(one, 0, 0);
        GLColor green = new GLColor(0, one, 0);
        GLColor blue = new GLColor(0, 0, one);
        GLColor yellow = new GLColor(one, one, 0);
        GLColor orange = new GLColor(one, half, 0);
        GLColor white = new GLColor(one, one, one);
        GLColor black = new GLColor(0, 0, 0);

        // coordinates for our cubes
        float c0 = -1.0f;
        float c1 = -0.38f;
        float c2 = -0.32f;
        float c3 = 0.32f;
        float c4 = 0.38f;
        float c5 = 1.0f;

        // top back, left to right
        mCubes[0]  = new Cube(world, c0, c4, c0, c1, c5, c1);
        mCubes[1]  = new Cube(world, c2, c4, c0, c3, c5, c1);
        mCubes[2]  = new Cube(world, c4, c4, c0, c5, c5, c1);
        // top middle, left to right
        mCubes[3]  = new Cube(world, c0, c4, c2, c1, c5, c3);
        mCubes[4]  = new Cube(world, c2, c4, c2, c3, c5, c3);
        mCubes[5]  = new Cube(world, c4, c4, c2, c5, c5, c3);
        // top front, left to right
        mCubes[6]  = new Cube(world, c0, c4, c4, c1, c5, c5);
        mCubes[7]  = new Cube(world, c2, c4, c4, c3, c5, c5);
        mCubes[8]  = new Cube(world, c4, c4, c4, c5, c5, c5);
        // middle back, left to right
        mCubes[9]  = new Cube(world, c0, c2, c0, c1, c3, c1);
        mCubes[10] = new Cube(world, c2, c2, c0, c3, c3, c1);
        mCubes[11] = new Cube(world, c4, c2, c0, c5, c3, c1);
        // middle middle, left to right
        mCubes[12] = new Cube(world, c0, c2, c2, c1, c3, c3);
        mCubes[13] = null;
        mCubes[14] = new Cube(world, c4, c2, c2, c5, c3, c3);
        // middle front, left to right
        mCubes[15] = new Cube(world, c0, c2, c4, c1, c3, c5);
        mCubes[16] = new Cube(world, c2, c2, c4, c3, c3, c5);
        mCubes[17] = new Cube(world, c4, c2, c4, c5, c3, c5);
        // bottom back, left to right
        mCubes[18] = new Cube(world, c0, c0, c0, c1, c1, c1);
        mCubes[19] = new Cube(world, c2, c0, c0, c3, c1, c1);
        mCubes[20] = new Cube(world, c4, c0, c0, c5, c1, c1);
        // bottom middle, left to right
        mCubes[21] = new Cube(world, c0, c0, c2, c1, c1, c3);
        mCubes[22] = new Cube(world, c2, c0, c2, c3, c1, c3);
        mCubes[23] = new Cube(world, c4, c0, c2, c5, c1, c3);
        // bottom front, left to right
        mCubes[24] = new Cube(world, c0, c0, c4, c1, c1, c5);
        mCubes[25] = new Cube(world, c2, c0, c4, c3, c1, c5);
        mCubes[26] = new Cube(world, c4, c0, c4, c5, c1, c5);

        // paint the sides
        int i, j;
        // set all faces black by default
        for (i = 0; i < 27; i++)
        {
            Cube cube = mCubes[i];
            if (cube != null)
            {
                for (j = 0; j < 6; j++)
                    cube.setFaceColor(j, black);
            }
        }

        // paint top
        for (i = 0; i < 9; i++)
            mCubes[i].setFaceColor(Cube.kTop, orange);
        // paint bottom
        for (i = 18; i < 27; i++)
            mCubes[i].setFaceColor(Cube.kBottom, red);
        // paint left
        for (i = 0; i < 27; i += 3)
            mCubes[i].setFaceColor(Cube.kLeft, yellow);
        // paint right
        for (i = 2; i < 27; i += 3)
            mCubes[i].setFaceColor(Cube.kRight, white);
        // paint back
        for (i = 0; i < 27; i += 9)
            for (j = 0; j < 3; j++)
                mCubes[i + j].setFaceColor(Cube.kBack, blue);
        // paint front
        for (i = 6; i < 27; i += 9)
            for (j = 0; j < 3; j++)
                mCubes[i + j].setFaceColor(Cube.kFront, green);

        for (i = 0; i < 27; i++)
            if (mCubes[i] != null)
                world.addShape(mCubes[i]);

        // initialize our permutation to solved position
        mPermutation = new int[27];
        for (i = 0; i < mPermutation.length; i++)
            mPermutation[i] = i;

        createLayers();
        updateLayers();

        world.generate();

        return world;
    }

    private void createLayers()
    {
        mLayers[kUp] = new Layer(Layer.kAxisY);
        mLayers[kDown] = new Layer(Layer.kAxisY);
        mLayers[kLeft] = new Layer(Layer.kAxisX);
        mLayers[kRight] = new Layer(Layer.kAxisX);
        mLayers[kFront] = new Layer(Layer.kAxisZ);
        mLayers[kBack] = new Layer(Layer.kAxisZ);
        mLayers[kMiddle] = new Layer(Layer.kAxisX);
        mLayers[kEquator] = new Layer(Layer.kAxisY);
        mLayers[kSide] = new Layer(Layer.kAxisZ);
    }

    private void updateLayers()
    {
        Layer layer;
        GLShape[] shapes;
        int i, j, k;

        // up layer
        layer = mLayers[kUp];
        shapes = layer.mShapes;
        for (i = 0; i < 9; i++)
            shapes[i] = mCubes[mPermutation[i]];

        // down layer
        layer = mLayers[kDown];
        shapes = layer.mShapes;
        for (i = 18, k = 0; i < 27; i++)
            shapes[k++] = mCubes[mPermutation[i]];

        // left layer
        layer = mLayers[kLeft];
        shapes = layer.mShapes;
        for (i = 0, k = 0; i < 27; i += 9)
            for (j = 0; j < 9; j += 3)
                shapes[k++] = mCubes[mPermutation[i + j]];

        // right layer
        layer = mLayers[kRight];
        shapes = layer.mShapes;
        for (i = 2, k = 0; i < 27; i += 9)
            for (j = 0; j < 9; j += 3)
                shapes[k++] = mCubes[mPermutation[i + j]];

        // front layer
        layer = mLayers[kFront];
        shapes = layer.mShapes;
        for (i = 6, k = 0; i < 27; i += 9)
            for (j = 0; j < 3; j++)
                shapes[k++] = mCubes[mPermutation[i + j]];

        // back layer
        layer = mLayers[kBack];
        shapes = layer.mShapes;
        for (i = 0, k = 0; i < 27; i += 9)
            for (j = 0; j < 3; j++)
                shapes[k++] = mCubes[mPermutation[i + j]];

        // middle layer
        layer = mLayers[kMiddle];
        shapes = layer.mShapes;
        for (i = 1, k = 0; i < 27; i += 9)
            for (j = 0; j < 9; j += 3)
                shapes[k++] = mCubes[mPermutation[i + j]];

        // equator layer
        layer = mLayers[kEquator];
        shapes = layer.mShapes;
        for (i = 9, k = 0; i < 18; i++)
            shapes[k++] = mCubes[mPermutation[i]];

        // side layer
        layer = mLayers[kSide];
        shapes = layer.mShapes;
        for (i = 3, k = 0; i < 27; i += 9)
            for (j = 0; j < 3; j++)
                shapes[k++] = mCubes[mPermutation[i + j]];
    }

    @Override
    protected void onCreate(Bundle savedInstanceState)
    {
        super.onCreate(savedInstanceState);

        // We don't need a title either.
        requestWindowFeature(Window.FEATURE_NO_TITLE);
        
        //Keep screen always on.
        getWindow().setFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON,
            WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON);
        
        mView = new GLSurfaceView(getApplication());
        mRenderer = new KubeRenderer(makeGLWorld(), this);
        mView.setRenderer(mRenderer);
        setContentView(mView);
    }

    @Override
    protected void onResume()
    {
        super.onResume();
        mView.onResume();
    }

    @Override
    protected void onPause()
    {
        super.onPause();
        mView.onPause();
    }

    public void animate()
    {
        // change our angle of view
        mRenderer.setAngle(mRenderer.getAngle() + 1.2f);

        if (mCurrentLayer == null)
        {
            int layerID = mRandom.nextInt(9);
            mCurrentLayer = mLayers[layerID];
            mCurrentLayerPermutation = mLayerPermutations[layerID];
            mCurrentLayer.startAnimation();
            
            boolean direction = mRandom.nextBoolean();
            int count = mRandom.nextInt(3) + 1;

            direction = false;
            count = 1;
            
            mCurrentAngle = 0;
            if (direction)
            {
                mAngleIncrement = (float)Math.PI / 50;
                mEndAngle = mCurrentAngle + ((float)Math.PI * count) / 2f;
            }
            else
            {
                mAngleIncrement = -(float)Math.PI / 50;
                mEndAngle = mCurrentAngle - ((float)Math.PI * count) / 2f;
            }
        }

        mCurrentAngle += mAngleIncrement;

        if ((mAngleIncrement > 0f && mCurrentAngle >= mEndAngle) ||
            (mAngleIncrement < 0f && mCurrentAngle <= mEndAngle))
        {
            mCurrentLayer.setAngle(mEndAngle);
            mCurrentLayer.endAnimation();
            mCurrentLayer = null;

            // adjust mPermutation based on the completed layer rotation
            int[] newPermutation = new int[27];
            for (int i = 0; i < 27; i++)
            {
                newPermutation[i] = mPermutation[mCurrentLayerPermutation[i]];
//    			newPermutation[i] = mCurrentLayerPermutation[mPermutation[i]];
            }
            mPermutation = newPermutation;
            updateLayers();
        }
        else
        {
            mCurrentLayer.setAngle(mCurrentAngle);
        }
    }

    GLSurfaceView mView;
    KubeRenderer mRenderer;
    Cube[] mCubes = new Cube[27];
    // a Layer for each possible move
    Layer[] mLayers = new Layer[9];
    // permutations corresponding to a pi/2 rotation of each layer about its axis
    static int[][] mLayerPermutations = {
        // permutation for UP layer
        { 2, 5, 8, 1, 4, 7, 0, 3, 6, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26 },
        // permutation for DOWN layer
        { 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 20, 23, 26, 19, 22, 25, 18, 21, 24 },
        // permutation for LEFT layer
        { 6, 1, 2, 15, 4, 5, 24, 7, 8, 3, 10, 11, 12, 13, 14, 21, 16, 17, 0, 19, 20, 9, 22, 23, 18, 25, 26 },
        // permutation for RIGHT layer
        { 0, 1, 8, 3, 4, 17, 6, 7, 26, 9, 10, 5, 12, 13, 14, 15, 16, 23, 18, 19, 2, 21, 22, 11, 24, 25, 20 },
        // permutation for FRONT layer
        { 0, 1, 2, 3, 4, 5, 24, 15, 6, 9, 10, 11, 12, 13, 14, 25, 16, 7, 18, 19, 20, 21, 22, 23, 26, 17, 8 },
        // permutation for BACK layer
        { 18, 9, 0, 3, 4, 5, 6, 7, 8, 19, 10, 1, 12, 13, 14, 15, 16, 17, 20, 11, 2, 21, 22, 23, 24, 25, 26 },
        // permutation for MIDDLE layer
        { 0, 7, 2, 3, 16, 5, 6, 25, 8, 9, 4, 11, 12, 13, 14, 15, 22, 17, 18, 1, 20, 21, 10, 23, 24, 19, 26 },
        // permutation for EQUATOR layer
        { 0, 1, 2, 3, 4, 5, 6, 7, 8, 11, 14, 17, 10, 13, 16, 9, 12, 15, 18, 19, 20, 21, 22, 23, 24, 25, 26 },
        // permutation for SIDE layer
        { 0, 1, 2, 21, 12, 3, 6, 7, 8, 9, 10, 11, 22, 13, 4, 15, 16, 17, 18, 19, 20, 23, 14, 5, 24, 25, 26 }
    };

    // current permutation of starting position
    int[] mPermutation;

    // for random cube movements
    Random mRandom = new Random(System.currentTimeMillis());
    // currently turning layer
    Layer mCurrentLayer = null;
    // current and final angle for current Layer animation
    float mCurrentAngle, mEndAngle;
    // amount to increment angle
    float mAngleIncrement;
    int[] mCurrentLayerPermutation;

    // names for our 9 layers (based on notation from http://www.cubefreak.net/notation.html)
    static final int kUp = 0;
    static final int kDown = 1;
    static final int kLeft = 2;
    static final int kRight = 3;
    static final int kFront = 4;
    static final int kBack = 5;
    static final int kMiddle = 6;
    static final int kEquator = 7;
    static final int kSide = 8;
}
